DAS/das/utils/logger.py

```diff
@@ -0,0 +1,235 @@
+    ----------
+    TOPICS: List(str)
+        List of topic names
+    _VERBOSE: bool
+        Specifies whether the incoming MQTT data and warnings are printed
+    _START_TIME: `time`
+        The current time used to produce time deltas
+    _LOG_FILE: `File`
+        Open log file object that can be written to
+    _LOG_FILE_WRITER: `csv.DictWriter`
+        Csv writter object that is used to write the data to the _LOG_FILE file
+    _client: `paho.mqtt.client`
+        MQTT client that connects to the broker and recives the messages
+    """
+
+    def __init__(self, csv_folder_path: str, topics: list = ["#"], broker_address: str = "localhost", verbose: bool = False) -> None:
+        # The logger object can subscribe to many topics (if none are selected then it will subscrive to all)
+        self.TOPICS = topics
+        self._VERBOSE = verbose
+        self._START_TIME = time.monotonic()
+
+        # Create csv_folder_path folder if none exists
+        Path(csv_folder_path).mkdir(parents=True, exist_ok=True)
+
+        # Name the csv log file xxxx_log.csv where xxxx is a number
+        previous_log_num = 0
+        for filename in os.listdir(csv_folder_path):
+            try:
+                if int(filename[0:4]) > previous_log_num:
+                    previous_log_num = abs(int(filename[0:4]))
+            except ValueError:
+                if self._VERBOSE:
+                    print(
+                        f"WARNING: {filename} should not be in {csv_folder_path}")
+            except Exception as e:
+                print(f"ERROR: {e}")
+
+        # fstring :0>4 used to ensure that the number will always be 4 long
+        filename = f"{previous_log_num + 1:0>4}_log.csv"
+        self._LOG_FILE = open(os.path.join(csv_folder_path, filename), "a")
+        self._LOG_FILE_WRITER = csv.DictWriter(
+            self._LOG_FILE,
+            delimiter=CsvConfig["delimiter"],
+            quotechar=CsvConfig["quotechar"],
+            quoting=CsvConfig["quoting"],
+            fieldnames=CsvConfig["fieldnames"],
+        )
+
+        # Add headers for csv
+        self._LOG_FILE_WRITER.writeheader()
+
+        # Connect to MQTT broker
+        self._client = mqtt.Client()
+        self._client.on_connect = self._on_connect
+        self._client.on_message = self._on_message
+        self._client.connect(broker_address)
+        self._client.loop_start()  # Threaded execution loop
+
+    def _on_connect(self, client, userdata, flags, rc):
+        """Callback function for MQTT broker on connection."""
+
+        if rc == 0:
+            print("Connection Successful!")
+        else:
+            raise ConnectionError(
+                "Connection was unsuccessful, check that the broker IP is corrrect")
+
+        # Subscribe to all of the topics
+        try:
+            for topic in self.TOPICS:
+                self._client.subscribe(topic)
+                if self._VERBOSE:
+                    print(f"Subscribed to: {topic}")
```

Comment:
Another possible future refactor for these scripts is to implement an actual logger so we don't have any `print()` statements and can properly display different levels of information.

Reply:
Just added the real python logger.

I looked into using it to do the real logging functionality, but its not designed to record data. It is only good at logging human readable output.

Reply:
Yeah i'm happy for it to be used as a way to log our debug statements for now!